File: java11/src/main/java/com/chilvers/henry/acer/Printer.java
package com.chilvers.henry.acer;

import de.vandermeer.asciitable.AT_Cell;
import de.vandermeer.asciitable.AsciiTable;
import de.vandermeer.asciitable.CWC_LongestLine;
import de.vandermeer.skb.interfaces.transformers.textformat.TextAlignment;
import java.util.LinkedList;

import static com.chilvers.henry.acer.ASCIIArtFontSize.ART_SIZE_MEDIUM;

public class Printer {
    private static final String WELCOME = "Welcome to...";
    private static final String APP_NAME = "A.C.E.R.";
    private static final String NUMBER_OF_PLAYERS_PROMPT = "Number of players? ";
    private static final String NUMBER_OF_DAYS_PROMPT = "Number of days in calendar? ";
    private static final String RESULTS = "Results...";
    private static final String READY_PROMPT = "Ready? Here we go... ";
    private static final String FAREWELL = "Thanks for using A.C.E.R.! Enjoy your advent calendar!";

    private final ASCIIArtGenerator artGen = new ASCIIArtGenerator();

    public void printOpeningScreen() {
        clearScreen();

        printWelcome();
        printLogo();
        printReady();

        printLogo();

    }

    public void printFinalResults(Game game) {
        clearScreen();
        printLogo();
        printResults();
        printPlayerTable(game);
        printFarewell();
    }

    public void clearScreen() {
        try {
            final String os = System.getProperty("os.name");

            if (os.contains("Windows")) {
                Runtime.getRuntime().exec("cls");
            } else {
                Runtime.getRuntime().exec("clear");
            }
        } catch (final Exception e) {
            //  Handle any exceptions.
        }
    }

    public void printNumberOfPlayersPrompt() {
        System.out.print(NUMBER_OF_PLAYERS_PROMPT);
    }

    public void printNumberOfDaysPrompt() {
        System.out.print(NUMBER_OF_DAYS_PROMPT);
    }

    public void printPlayerNamePrompt(Integer i) {
        System.out.printf("Name of Player %d? ", i);
    }

    public void printPlayerTurnResult(String playerName, Integer dayToOpen) {
        System.out.println(playerName + ", you got day " + dayToOpen);
        System.out.println();
    }

    void printPlayerTable(Game game) {
        var at = new AsciiTable();
        var cwc = new CWC_LongestLine();

        at.addRule();
        var headerCells = at.addRow("Player", "Days to Open").getCells();
        at.addRule();
        setCellPadding(headerCells);
        headerCells.get(1).getContext().setTextAlignment(TextAlignment.CENTER);

        game.getPlayers().forEach(it -> {
            var cells = at.addRow(it.getName(), it.finalResults()).getCells();
            setCellPadding(cells);
        });
        at.addRule();

        at.getRenderer().setCWC(cwc);
        System.out.println(at.render());
    }

    private void setCellPadding(LinkedList<AT_Cell> cells) {
        cells.get(0).getContext().setPaddingLeftChar(' ');
        cells.get(0).getContext().setPaddingRightChar(' ');
        cells.get(0).getContext().setPaddingLeft(1);
        cells.get(0).getContext().setPaddingRight(1);

        cells.get(1).getContext().setPaddingLeftChar(' ');
        cells.get(1).getContext().setPaddingRightChar(' ');
        cells.get(1).getContext().setPaddingLeft(1);
        cells.get(1).getContext().setPaddingRight(1);
    }

    private void printFarewell() {
        System.out.println();
        System.out.println(FAREWELL);
    }

    private void printWelcome() {
        System.out.println(WELCOME);
    }

    private void printResults() {
        System.out.println(RESULTS);
    }

    private void printLogo() {
        System.out.println();
        artGen.printTextArt(APP_NAME, ART_SIZE_MEDIUM);
        System.out.println("( Advent           Calendar          Enjoyment        Randomizer )");
        System.out.println();
    }

    public void printDaysRemaining(Integer daysRemaining) {
        System.out.println("[Days remaining: " + daysRemaining + "]");
    }

    public void printRequestPlayerInteraction(String currentPlayerName) {
        System.out.println(currentPlayerName + ", hit <enter> to pick a number...");
    }

    private void printReady() {
        System.out.println(READY_PROMPT);
    }
}

File: kotlin/src/main/kotlin/com/chilvers/henry/acer/Printer.kt
package com.chilvers.henry.acer

import de.vandermeer.asciitable.AT_Cell
import de.vandermeer.asciitable.AsciiTable
import de.vandermeer.asciitable.CWC_LongestLine
import de.vandermeer.skb.interfaces.transformers.textformat.TextAlignment
import java.util.*

class Printer {
    private val artGen: ASCIIArtGenerator = ASCIIArtGenerator()

    @Throws(InterruptedException::class)
    fun printOpeningScreen() {
        clearScreen()

        printWelcome()
        printLogo()
        printReady()
        printLogo()
    }

    fun printFinalResults(game: Game) {
        clearScreen()
        printLogo()
        printResults()
        printPlayerTable(game)
        printFarewell()
    }

    private fun clearScreen() {
        try {
            val os = System.getProperty("os.name")

            if (os.contains("Windows")) {
                Runtime.getRuntime().exec("cls")
            } else {
                Runtime.getRuntime().exec("clear")
            }
        } catch (e: Exception) {
            //  Handle any exceptions.
        }
    }

    fun printNumberOfPlayersPrompt() {
        print(NUMBER_OF_PLAYERS_PROMPT)
    }

    fun printNumberOfDaysPrompt() {
        print(NUMBER_OF_DAYS_PROMPT)
    }

    fun printPlayerNamePrompt(i: Int?) {
        System.out.printf("Name of Player %d? ", i)
    }

    fun printPlayerTurnResult(playerName: String, dayToOpen: Int) {
        println("$playerName, you got day $dayToOpen")
        println()
    }

    fun printPlayerTable(game: Game) {
//        val longestName = max(game.getLongestPlayersName(), "Player".length) + 1
        val at = AsciiTable()
        val cwc = CWC_LongestLine()

        at.addRule()
        val headerCells = at.addRow("Player", "Days to Open").cells
        at.addRule()
        setCellPadding(headerCells)
        headerCells[1].context.textAlignment = TextAlignment.CENTER

        game.players.forEach {
            val cells = at.addRow(it.name, it.finalResults()).cells
            setCellPadding(cells)
        }
        at.addRule()

        at.renderer.cwc = cwc
        println(at.render()) //todo: Calculate max width??
    }

    fun printDaysRemaining(daysRemaining: Int) {
        println("[Days remaining: $daysRemaining]")
    }

    fun printRequestPlayerInteraction(currentPlayerName: String) {
        println("$currentPlayerName, hit <enter> to pick a number...")
    }

    private fun printFarewell() {
        println()
        println(FAREWELL)
    }

    private fun printWelcome() {
        println(WELCOME)
    }

    private fun printResults() {
        println(RESULTS)
    }

    private fun printLogo() {
        println()
        artGen.printTextArt(APP_NAME, ASCIIArtFontSize.ART_SIZE_MEDIUM)
        println("( Advent           Calendar          Enjoyment        Randomizer )")
        println()
    }

    private fun printReady() {
        println(READY_PROMPT)
    }

    private fun setCellPadding(cells: LinkedList<AT_Cell>) {
        cells[0].context.paddingLeftChar = ' '
        cells[0].context.paddingRightChar = ' '
        cells[0].context.paddingLeft = 1
        cells[0].context.paddingRight = 1

        cells[1].context.paddingLeftChar = ' '
        cells[1].context.paddingRightChar = ' '
        cells[1].context.paddingLeft = 1
        cells[1].context.paddingRight = 1
    }

    companion object {
        private const val WELCOME = "Welcome to..."
        private const val APP_NAME = "A.C.E.R."
        private const val NUMBER_OF_PLAYERS_PROMPT = "Number of players? "
        private const val NUMBER_OF_DAYS_PROMPT = "Number of days in calendar? "
        private const val RESULTS = "Results..."
        private const val READY_PROMPT = "Ready? Here we go... "
        private const val FAREWELL = "Thanks for using A.C.E.R.! Enjoy your advent calendar!"
    }
}